From 4541d103c6dd0f38251b38e2c81c3101a9f2ad17 Mon Sep 17 00:00:00 2001
From: Dmitry Neverov <dmitry.neverov@jetbrains.com>
Date: Sun, 14 Aug 2022 09:38:13 +0200
Subject: [PATCH] [dbg] track variable writes only for builtin commands


diff --git a/Source/Sysprogs/HLDPServer.cpp b/Source/Sysprogs/HLDPServer.cpp
index 76ba802223..80aabadae0 100644
--- a/Source/Sysprogs/HLDPServer.cpp
+++ b/Source/Sysprogs/HLDPServer.cpp
@@ -377,6 +377,16 @@ namespace Sysprogs
 		if (m_Detached)
 			return nullptr;
 
+		if (!m_VarWriteCommandsInitialized) {
+			std::vector<std::string> commands = pMakefile->GetState()->GetCommandNames();
+			for (auto it : commands) {
+				m_VarWriteCommands.insert(it);
+			}
+			// exclude project as it writes a lot of internal variables
+			m_VarWriteCommands.erase("project");
+			m_VarWriteCommandsInitialized = true;
+		}
+
 		std::unique_ptr<RAIIScope> pScope(new RAIIScope(this, pMakefile, function));
 		struct
 		{
@@ -506,8 +516,11 @@ namespace Sysprogs
 		{
 			if (m_CallStack.size() > 0 && !IsIgnoreVarWrite(variable))
 			{
-				const cmListFileContext &topFile = mf->GetBacktrace().Top();
-				m_VarWrites[topFile.FilePath][variable] = std::make_pair(std::string(newValue), topFile.Line);
+				std::string const& funcName = m_CallStack[m_CallStack.size()-1]->Function.OriginalName();
+				if (m_VarWriteCommands.find(funcName) != m_VarWriteCommands.end()) {
+					const cmListFileContext &topFile = mf->GetBacktrace().Top();
+					m_VarWrites[topFile.FilePath][variable] = std::make_pair(std::string(newValue), topFile.Line);
+				}
 			}
 		}
 
diff --git a/Source/Sysprogs/HLDPServer.h b/Source/Sysprogs/HLDPServer.h
index cc891a5839..a7296038e1 100644
--- a/Source/Sysprogs/HLDPServer.h
+++ b/Source/Sysprogs/HLDPServer.h
@@ -123,6 +123,12 @@ namespace Sysprogs
 		// file path -> var name -> var value and the line of the last write access
 		std::unordered_map<std::string, std::unordered_map<std::string, std::pair<std::string, int>>> m_VarWrites;
 
+		// Set of command names for which we want to track variable writes.
+		// Includes only builtin commands to filter out writes to parameters
+		// of user-defined functions.
+		std::unordered_set<std::string> m_VarWriteCommands;
+		bool m_VarWriteCommandsInitialized = false;
+
 		// Holds names of all created targets
 		std::set<std::string> m_TargetNames;
 		enum
-- 
2.50.1 (Apple Git-155)

