From c287591b0d9b3130240f7b88d5cd016133ec1f4f Mon Sep 17 00:00:00 2001
From: "Dmitry.Neverov" <Dmitry.Neverov@jetbrains.com>
Date: Fri, 28 May 2021 16:18:47 -0700
Subject: [PATCH] [dbg] make it compile


diff --git a/Source/Sysprogs/HLDPServer.cpp b/Source/Sysprogs/HLDPServer.cpp
index bd1d62e3f1..5bf76bea67 100644
--- a/Source/Sysprogs/HLDPServer.cpp
+++ b/Source/Sysprogs/HLDPServer.cpp
@@ -239,7 +239,7 @@ namespace Sysprogs
 			std::string empty;
 			for (const std::string &key : properties.GetKeys())
 			{
-				cmProp value = properties.GetPropertyValue(key);
+				cmValue value = properties.GetPropertyValue(key);
 				result.push_back(std::make_unique<SimpleExpression>(key, "(property entry)", value ? *value : empty));
 			}
 			return std::move(result);
@@ -572,14 +572,14 @@ namespace Sysprogs
 		if (!m_pSocket->Write(&hdr, sizeof(hdr)))
 		{
 			cmSystemTools::Error("Failed to write debug protocol reply header.");
-			cmSystemTools::SetFatalErrorOccured();
+			cmSystemTools::SetFatalErrorOccurred();
 			return false;
 		}
 
 		if (!m_pSocket->Write(builder.GetBuffer().data(), hdr.PayloadSize))
 		{
 			cmSystemTools::Error("Failed to write debug protocol reply payload.");
-			cmSystemTools::SetFatalErrorOccured();
+			cmSystemTools::SetFatalErrorOccurred();
 			return false;
 		}
 
@@ -592,7 +592,7 @@ namespace Sysprogs
 		if (!m_pSocket->ReadAll(&hdr, sizeof(hdr)))
 		{
 			cmSystemTools::Error("Failed to receive debug protocol request header.");
-			cmSystemTools::SetFatalErrorOccured();
+			cmSystemTools::SetFatalErrorOccurred();
 			return HLDPPacketType::Invalid;
 		}
 
@@ -602,7 +602,7 @@ namespace Sysprogs
 			if (!m_pSocket->ReadAll(pBuffer, hdr.PayloadSize))
 			{
 				cmSystemTools::Error("Failed to receive debug protocol request payload.");
-				cmSystemTools::SetFatalErrorOccured();
+				cmSystemTools::SetFatalErrorOccurred();
 				return HLDPPacketType::Invalid;
 			}
 		}
@@ -733,7 +733,7 @@ namespace Sysprogs
 				return;
 			case HLDPPacketType::csTerminate:
 				cmSystemTools::Error("Configuration aborted via debugging interface.");
-				cmSystemTools::SetFatalErrorOccured();
+				cmSystemTools::SetFatalErrorOccurred();
 				return;
 			case HLDPPacketType::csCreateExpression:
 			{
@@ -866,7 +866,7 @@ namespace Sysprogs
 		if (m_pServer->m_CallStack.back() != this)
 		{
 			cmSystemTools::Error("CMake scope imbalance detected");
-			cmSystemTools::SetFatalErrorOccured();
+			cmSystemTools::SetFatalErrorOccurred();
 		}
 
 		if (m_UniqueID == m_pServer->m_EndOfStepScopeID)
@@ -891,7 +891,7 @@ namespace Sysprogs
 				return nullptr;
 		}
 
-		const std::string *pValue = cmDefinitions::Get(text, scope.Position->Vars, scope.Position->Root);
+		cmValue pValue = cmDefinitions::Get(text, scope.Position->Vars, scope.Position->Root);
 		if (pValue)
 			return std::make_unique<VariableExpression>(scope, text, pValue->c_str());
 
@@ -899,7 +899,7 @@ namespace Sysprogs
 		if (pTarget)
 			return std::make_unique<TargetExpression>(pTarget);
 
-		cmProp pCacheValue = scope.Makefile->GetState()->GetCacheEntryValue(text);
+		cmValue pCacheValue = scope.Makefile->GetState()->GetCacheEntryValue(text);
 		if (pCacheValue)
 		{
 			return std::make_unique<CacheEntryExpression>(text, pCacheValue->c_str());
diff --git a/Source/Sysprogs/HLDPServer.h b/Source/Sysprogs/HLDPServer.h
index 5041b28a6c..a5729e0f09 100644
--- a/Source/Sysprogs/HLDPServer.h
+++ b/Source/Sysprogs/HLDPServer.h
@@ -6,7 +6,7 @@
 
 class cmCommand;
 class BasicIncomingSocket;
-struct cmListFileFunction;
+class cmListFileFunction;
 enum class MessageType;
 
 namespace Sysprogs
diff --git a/Source/cmake.cxx b/Source/cmake.cxx
index 36e7d3156b..3519acdf11 100644
--- a/Source/cmake.cxx
+++ b/Source/cmake.cxx
@@ -4409,7 +4409,7 @@ void cmake::StartDebugServerIfEnabled()
     m_pDebugServer.reset(new Sysprogs::HLDPServer(DebugServerPort));
     if (!m_pDebugServer->WaitForClient()) {
       cmSystemTools::Error("Failed to start debugging server. Aborting...");
-      cmSystemTools::SetFatalErrorOccured();
+      cmSystemTools::SetFatalErrorOccurred();
     }
   }
 }
-- 
2.50.1 (Apple Git-155)

