From 1a4e4ffaa47e3a84e6fc68afcd23558a85e9c03a Mon Sep 17 00:00:00 2001
From: "Dmitry.Neverov" <dmitry.neverov@jetbrains.com>
Date: Mon, 8 Aug 2022 17:04:17 +0200
Subject: [PATCH] [dbg] ability to get all variable writes in a file


diff --git a/Source/Sysprogs/HLDP.h b/Source/Sysprogs/HLDP.h
index b2fb3cd410..a7cb44d564 100644
--- a/Source/Sysprogs/HLDP.h
+++ b/Source/Sysprogs/HLDP.h
@@ -60,6 +60,9 @@ namespace Sysprogs
 
 		scDebugMessage, // Payload: Stream:int32, text:string. Stream is implementation-specific.
 		scTargetExited, // Payload: exit code
+
+		csVarsIn,  // frameId:int32, one-based fromLine:int32 (inclusive), one-based toLine:int32 (exclusive)
+		scVarsOut, // array of [var name:string, var value:string, line]
 	};
 
 	enum class TargetStopReason
diff --git a/Source/Sysprogs/HLDPServer.cpp b/Source/Sysprogs/HLDPServer.cpp
index 286e15120f..76ba802223 100644
--- a/Source/Sysprogs/HLDPServer.cpp
+++ b/Source/Sysprogs/HLDPServer.cpp
@@ -500,11 +500,20 @@ namespace Sysprogs
 
 	void HLDPServer::OnVariableAccessed(const std::string &variable, int access_type, const char *newValue, const cmMakefile *mf)
 	{
+		bool isRead = access_type == cmVariableWatch::VARIABLE_READ_ACCESS || access_type == cmVariableWatch::UNKNOWN_VARIABLE_READ_ACCESS;
+
+		if (!isRead && newValue)
+		{
+			if (m_CallStack.size() > 0 && !IsIgnoreVarWrite(variable))
+			{
+				const cmListFileContext &topFile = mf->GetBacktrace().Top();
+				m_VarWrites[topFile.FilePath][variable] = std::make_pair(std::string(newValue), topFile.Line);
+			}
+		}
+
 		if (m_WatchedVariables.find(variable) == m_WatchedVariables.end())
 			return;
 
-		bool isRead = access_type == cmVariableWatch::VARIABLE_READ_ACCESS || access_type == cmVariableWatch::UNKNOWN_VARIABLE_READ_ACCESS;
-
 		auto id = m_BreakpointManager.TryLocateEnabledDomainSpecificBreakpoint([&](BasicBreakpointManager::DomainSpecificBreakpointExtension *pBp) {
 			switch (static_cast<DomainSpecificBreakpoint *>(pBp)->Type)
 			{
@@ -523,6 +532,32 @@ namespace Sysprogs
 			ReportStopAndServeDebugRequests(TargetStopReason::Breakpoint, id, "", nullptr);
 	}
 
+	bool HLDPServer::IsIgnoreVarWrite(const std::string &variable)
+	{
+		if (variable == "ARGC") {
+			return true;
+		}
+		if (variable == "ARGN") {
+			return true;
+		}
+		if (variable == "CMAKE_CURRENT_FUNCTION") {
+			return true;
+		}
+		if (variable == "CMAKE_CURRENT_FUNCTION_LIST_DIR") {
+			return true;
+		}
+		if (variable == "CMAKE_CURRENT_FUNCTION_LIST_FILE") {
+			return true;
+		}
+		if (variable == "CMAKE_CURRENT_FUNCTION_LIST_LINE") {
+			return true;
+		}
+		if (variable.rfind("ARGV", 0) == 0) {
+			return true;
+		}
+		return false;
+	}
+
 	void HLDPServer::OnTargetCreated(cmStateEnums::TargetType type, const std::string &targetName)
 	{
 		if (!m_CallStack.empty()) // some auxiliary targets are created with empty stack, ignore them since we cannot evaluate them
@@ -930,6 +965,40 @@ namespace Sysprogs
 						SendErrorPacket(error);
 				}
 				break;
+			case HLDPPacketType::csVarsIn:
+			{
+				int fromLine = 0;
+				int toLine = 0;
+
+				if (!reader.ReadInt32(&ID) || !reader.ReadInt32(&fromLine) || !reader.ReadInt32(&toLine))
+				{
+					SendErrorPacket("Invalid expression request");
+					continue;
+				}
+				RAIIScope *frame = 0 <= ID && ID < m_CallStack.size() ? m_CallStack[ID] : nullptr;
+				if (!frame)
+				{
+					SendErrorPacket("Invalid frame ID");
+				}
+				else
+				{
+					std::unordered_map<std::string, std::pair<std::string, int>> fileMap = m_VarWrites[frame->SourceFile];
+					auto varCount = builder.AppendDelayedInt32();
+					for (auto& it: fileMap) {
+						auto varName = it.first;
+						auto varValueAndLine = it.second;
+						int line = varValueAndLine.second;
+						if (fromLine <= line && line < toLine)
+						{
+							builder.AppendString(varName);
+							builder.AppendString(varValueAndLine.first);
+							builder.AppendInt32(line);
+							(*varCount)++;
+						}
+					}
+					SendReply(HLDPPacketType::scVarsOut, builder);
+				}
+			} break;
 			default:
 				if (requestType > HLDPPacketType::BeforeFirstBreakpointRelatedCommand && requestType < HLDPPacketType::AfterLastBreakpointRelatedCommand)
 				{
diff --git a/Source/Sysprogs/HLDPServer.h b/Source/Sysprogs/HLDPServer.h
index f3ad27ed2a..cc891a5839 100644
--- a/Source/Sysprogs/HLDPServer.h
+++ b/Source/Sysprogs/HLDPServer.h
@@ -1,7 +1,9 @@
 #pragma once
 #include <map>
+#include <unordered_map>
 #include <string>
 #include <vector>
+#include <utility>
 #include "BasicBreakpointManager.h"
 
 class cmCommand;
@@ -63,6 +65,7 @@ namespace Sysprogs
 		void SendErrorPacket(std::string details);
 
 		void ReportStopAndServeDebugRequests(TargetStopReason stopReason, unsigned intParam, const std::string &stringParam, bool *pSkipThisInstruction);
+		bool IsIgnoreVarWrite(const std::string &variable);
 
 	private:
 		class ExpressionBase
@@ -117,6 +120,9 @@ namespace Sysprogs
 		//Set of variables that ever had watches created. This should reduce the delay when checking each variable access.
 		std::set<BasicBreakpointManager::CaseInsensitiveObjectName> m_WatchedVariables;
 
+		// file path -> var name -> var value and the line of the last write access
+		std::unordered_map<std::string, std::unordered_map<std::string, std::pair<std::string, int>>> m_VarWrites;
+
 		// Holds names of all created targets
 		std::set<std::string> m_TargetNames;
 		enum
-- 
2.50.1 (Apple Git-155)

